### PR TITLE
Feature/login backend

### LINKED ISSUE #67: CI Pipeline `Unzip Backend Artifact` step failed
Runs normally then gives an error exit code without much indication as to what happened.
https://github.com/UniversityOfSaskatchewanCMPT371/term-project-2024-team-4/actions/runs/8088746225/job/22103431159?pr=65
image

### PR DESCRIPTION
**Pull request by DcGitLoom**

### CODE REVIEW

File: backend/.env
Comment (lecegues) [inline backend/.env:1]: This file should not be here (security risk)

File: backend/helpers/register.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+const { User } = require("../dist/user.entity");
+const dataSource = require("../config/db");
+const bcrypt = require("bcrypt");
+
+const registerUser = async () => {
```
Comment (lecegues) [inline backend/helpers/register.js:17]: Include API Docs (pre-conditions, purpose, post-conditions)

**Review by lecegues — CHANGES_REQUESTED** [review]
Needs data validation (e.g. spaces, special characters, etc...), use of `logger` not `console.log` and more API Docs.
Additionally, the feature has no protection from multiple failed attempts (e.g. someone attempts to log in **wrongly** multiple times; should handle using lockout, captcha, etc.); this will protect from brute force attacks.

File: backend/routes/users.js [hunk 2]
```diff
@@ -1,9 +1,129 @@
...
-	res.send("respond with a resource");
+// Function to generate a secure random string
+function generateJWTSecret(length = 64) {
+	return crypto.randomBytes(length).toString("hex");
+}
+
+// Generate a JWT secret
+const JWT_SECRET = generateJWTSecret();
+
+router.post("/", async (req, res) => {
```
Comment (lecegues) [inline backend/routes/users.js:33]: API Docs

File: backend/routes/users.js [hunk 3]
```diff
@@ -1,9 +1,129 @@
...
+		} else {
+			return res.status(401).json({ message: "Unauthorized" });
+		}
+	} catch (error) {
+		console.error("Error creating user:", error);
+		res.status(500).json({ message: "Internal server error" });
+	}
+});
+
+router.post("/logout", async (req, res) => {
```
Comment (lecegues) [inline backend/routes/users.js:100]: API Docs

File: backend/routes/users.js
Comment (lecegues) [inline backend/routes/users.js:1]: Needs field validation (e.g. right now, spaces will work in usernames and paswords ["hello world"] and should not)

Comment (lecegues) [inline backend/routes/users.js:1]: Use `logger` middleware rather than `console.log` as it will send logs to log files and have color-coded logs.

**Review by lecegues — CHANGES_REQUESTED** [review]

File: frontend/src/components/LoginModal.jsx [hunk 4]
```diff
@@ -31,33 +21,57 @@ function LoginModal({ modalVisible, closeModal }) {
 	const handleSubmit = async (event) => {
 		event.preventDefault();
 
-		//**Not needed yet */
-		// const response = await loginUser({
-		//   username,
-		//   password
-		// });
+		if (userName.length < 5) {
```
Comment (lecegues) [inline frontend/src/components/LoginModal.jsx]: Sorry, Miscommunication upon the required validation feature. **Username and password validation** should be present upon signup/changing username or password. You have changed the `patch` endpoint already, but the change in the `login-modal` is unnecessary

**Review by lecegues — CHANGES_REQUESTED** [review]

File: backend/routes/users.js [hunk 5]
```diff
@@ -1,9 +1,252 @@
...
+ * - If an error occurs during token verification, returns a message "Internal server error" with status code 500.
+ */
+// GET route to fetch the single user's details
+router.get("/", async (req, res) => {
+	const { token } = req.cookies;
+	if (token) {
+		jwt.verify(token, JWT_SECRET, {}, (err, user) => {
+			if (err) {
+				logger.error("Error verifying JWT:", err);
+				throw err;
```
Comment (lecegues) [inline backend/routes/users.js:136]: This throws an error (which effectively crashes the whole backend) if the user does not have a token. Please handle this more gracefully.

**Review by lecegues — CHANGES_REQUESTED** [review]

File: backend/routes/users.js (hunk 5 above)
Comment (lecegues) [reply backend/routes/users.js:136]: I don't know if this is where the exact error originated from, but a more detailed outlook of the issue is issue #92

Comment (lecegues) [reply backend/routes/users.js:136]: Due to the need for this branch to allow developers to start development for ID4, we will leave this issue for a future patch. (issue #92)

**Review by funmise — APPROVED** [review]
No further comments besides the ones John has already stated. Issue with persistent cookies is still present but I advise merging for now and assigning a fix to new Dev

**Review by lecegues — APPROVED** [review]

**Pull request merged**